File: leet-code/332. Reconstruct Itinerary/ReconstructItinerary332.java
import java.util.*;

public class ReconstructItinerary332 {
    public List<String> findItinerary(List<List<String>> tickets) {
        Map<String, PriorityQueue<String>> map = new HashMap<>();
        for (List<String> ticket : tickets) {
            map.putIfAbsent(ticket.get(0), new PriorityQueue<>());
            map.get(ticket.get(0)).add(ticket.get(1));
        }
        return dfs(map, new ArrayList<>(), "JFK");
    }

    private List<String> dfs(Map<String, PriorityQueue<String>> map,
                             List<String> itinerary,
                             String departure) {
        while (map.containsKey(departure) && !map.get(departure).isEmpty()) {
            String destination = map.get(departure).poll();
            dfs(map, itinerary, destination);
        }
        itinerary.add(0, departure);
        return itinerary;
    }
}

//[
// ["ANU","EZE"],
// ["EZE","AXA"],
// ["AXA","TIA"],
// ["TIA","ANU"],
// ["ANU","JFK"],
// ["JFK","ANU"],
// ["TIA","ANU"],
// ["TIA","JFK"],
// ["ANU","TIA"],
// ["JFK","TIA"],
//]

File: leet-code/332. Reconstruct Itinerary/ReconstructItinerary332.kt
package com.example

import java.util.PriorityQueue

class ReconstructItinerary332 {
    fun findItinerary(tickets: List<List<String>>): List<String> {
        val map = mutableMapOf<String, PriorityQueue<String>>()
        tickets.forEach { ticket ->
            val start = ticket[0]
            val end = ticket[1]
            map.putIfAbsent(start, PriorityQueue())
            map[ticket[0]]!!.add(end)
        }
        val itinerary = mutableListOf<String>()
        dfs(map, itinerary, "JFK")
        return itinerary
    }

    private fun dfs(
        map: MutableMap<String, PriorityQueue<String>>,
        itinerary: MutableList<String>,
        departure: String,
    ) {
        while (map.contains(departure) && map[departure]!!.isNotEmpty()) {
            dfs(map, itinerary, map[departure]!!.poll())
        }
        itinerary.add(0, departure)
    }
}
